File: android/app/src/main/kotlin/com/inferenceaftermath/remotly/ui/ApprovalCard.kt
// Shown while the watched pane is `blocked` (DESIGN.md §4.5): the dialog as a card — tool name, command or file,
// the agent's one-line summary, its question — and the choices it offers as buttons in the dialog's own words.
// Tapping one asks the bridge to move the desktop cursor there and press Enter (`choose`), so the same card serves
// permission prompts, Claude's AskUserQuestion menus and pickers; the choice the desktop currently marks is
// highlighted. Without parsed details (trust dialogs, unknown layouts) the card falls back to the agent's key-map
// actions: Approve / Approve for session / Deny / Interrupt, and the quiet "Deny with feedback…".
package com.inferenceaftermath.remotly.ui

import androidx.compose.animation.Crossfade
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.inferenceaftermath.remotly.core.protocol.ApprovalAction
import com.inferenceaftermath.remotly.core.protocol.ApprovalResult
import com.inferenceaftermath.remotly.core.protocol.Outcome
import com.inferenceaftermath.remotly.core.protocol.PaneInfo

@Composable
fun ApprovalCard(
    pane: PaneInfo,
    result: ApprovalResult?,
    busy: Boolean,
    onChoose: (option: Int, label: String) -> Unit,
    onAction: (action: String, feedback: String?, force: Boolean) -> Unit,
) {
    var showFeedback by remember { mutableStateOf(false) }
    var lastAction by remember(pane.prompt_id) { mutableStateOf<Pair<String, String?>?>(null) }
    fun act(action: String, feedback: String? = null) {
        lastAction = action to feedback
        onAction(action, feedback, false)
    }
    val details = pane.approval
    val isChoice = details?.isChoice == true
    val sent = result != null && result.outcome == Outcome.SENT && !busy
    val failed = result != null && result.outcome != Outcome.SENT && !busy
    val toolLabel = details?.tool ?: if (isChoice) "Question" else "Approval"
    val kindLabel = when {
        details == null -> "prompt"
        isChoice -> "question"
        else -> "permission"
    }
    val quiet = if (isChoice) "Something else…" else "Deny with feedback…"

    Column(Modifier.fillMaxWidth().background(Tokens.panel)) {
        Hairline()
        Crossfade(targetState = sent, animationSpec = tween(200), label = "approval") { isSent ->
            Column(Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 12.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                val r = result
                if (isSent && r != null) {
                    HeadRow("✓ Answered from your phone", Tokens.done, "sent")
                    Text(describe(r), style = Type.description)
                } else {
                    when {
                        busy -> HeadRow(toolLabel, Tokens.accent, "sending…")
                        failed && r != null -> HeadRow("✕ Not sent", Tokens.blocked, r.outcome.replace('_', ' '))
                        else -> HeadRow(toolLabel, Tokens.accent, kindLabel)
                    }
                    if (failed && r != null) Text(describe(r), style = Type.description)
                    if (details != null) {
                        val what = details.command ?: details.path
                        if (what != null) CommandBlock(what)
                        details.description?.takeIf { it.isNotBlank() && it != what }?.let { Text(it, style = Type.description) }
                        if (details.question.isNotEmpty()) Text(details.question, style = Type.question)
                        if (details.options.isNotEmpty()) {
                            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                                details.options.forEachIndexed { index, option ->
                                    val number = index + 1
                                    val marked = details.selected == number
                                    OptionButton(
                                        option, marked = marked, enabled = !busy,
                                        onClick = { onChoose(number, option) },
                                        description = "Option $number: $option" + if (marked) ", marked on the desktop" else "",
                                    )
                                }
                            }
                        }
                    } else {
                        pane.state_label?.takeIf { it.isNotBlank() }?.let { Text(it, style = Type.description) }
                        Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                            OptionButton("Approve", marked = true, enabled = !busy, onClick = { act(ApprovalAction.APPROVE) })
                            OptionButton("Approve for session", marked = false, enabled = !busy, onClick = { act(ApprovalAction.APPROVE_SESSION) })
                            OptionButton("Deny", marked = false, enabled = !busy, onClick = { act(ApprovalAction.DENY) })
                            OptionButton("Interrupt", marked = false, enabled = !busy, onClick = { act(ApprovalAction.INTERRUPT) })
                        }
                    }
                    Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                        QuietAction(quiet, enabled = !busy) { showFeedback = true }
                        val last = lastAction
                        if (failed && r?.outcome == Outcome.SIGNATURE_MISMATCH && last != null) {
                            QuietAction("Send anyway", enabled = !busy) { onAction(last.first, last.second, true) }
                        }
                    }
                }
            }
        }
    }
    if (showFeedback) {
        FeedbackDialog(
            title = quiet,
            onDismiss = { showFeedback = false },
            onSend = { text ->
                showFeedback = false
                act(ApprovalAction.DENY_FEEDBACK, text.ifBlank { null })
            },
        )
    }
}

@Composable
private fun HeadRow(left: String, leftColor: Color, right: String) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Text(left, style = Type.cardHead.copy(fontWeight = FontWeight.Bold), color = leftColor, maxLines = 1, overflow = TextOverflow.Ellipsis, modifier = Modifier.weight(1f))
        Text(right, style = Type.cardHead, color = Tokens.fg3, maxLines = 1, overflow = TextOverflow.Ellipsis, modifier = Modifier.padding(start = 8.dp))
    }
}

/** What runs: mono on `bg`, `line` border, radius 8, up to 6 lines, selectable. */
@Composable
private fun CommandBlock(text: String) {
    val shape = RoundedCornerShape(8.dp)
    SelectionContainer {
        Text(
            text,
            style = Type.command,
            maxLines = 6,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.fillMaxWidth().clip(shape).background(Tokens.bg).border(1.dp, Tokens.line, shape).padding(horizontal = 8.dp, vertical = 6.dp),
        )
    }
}

@Composable
private fun QuietAction(text: String, enabled: Boolean, onClick: () -> Unit) {
    Text(
        text,
        style = Type.small,
        color = Tokens.fg2,
        modifier = Modifier.alpha(if (enabled) 1f else 0.6f).clip(RoundedCornerShape(6.dp)).clickable(enabled = enabled, onClick = onClick).padding(vertical = 4.dp),
    )
}

/** "Deny with feedback…" / "Something else…": Esc first, then the words are typed to the agent. */
@Composable
private fun FeedbackDialog(title: String, onDismiss: () -> Unit, onSend: (String) -> Unit) {
    var feedback by remember { mutableStateOf("") }
    AlertDialog(
        onDismissRequest = onDismiss,
        containerColor = Tokens.panel,
        titleContentColor = Tokens.fg,
        textContentColor = Tokens.fg2,
        shape = RoundedCornerShape(16.dp),
        title = { Text(title, style = Type.navTitle) },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                FlowField(feedback, { feedback = it }, placeholder = "What should it do instead?", singleLine = false, minLines = 3, background = Tokens.panel2)
                Text("Esc dismisses the dialog first, then your words are typed to the agent.", style = Type.small.copy(fontSize = 13.sp), color = Tokens.fg3)
            }
        },
        confirmButton = {
            // Blank feedback would fall back to a plain deny; the button waits for words (same on iOS).
            val ready = feedback.isNotBlank()
            TextButton(enabled = ready, onClick = { onSend(feedback) }) { Text("Send", color = if (ready) Tokens.interactive else Tokens.fg3, fontWeight = FontWeight.SemiBold) }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text("Cancel", color = Tokens.fg2) } },
    )
}

fun describe(r: ApprovalResult): String = when (r.outcome) {
    Outcome.SENT -> "Sent" + (r.status_after?.let { " · agent now $it" } ?: "")
    Outcome.STALE -> "Stale: the prompt changed before the keys were sent"
    Outcome.NOT_BLOCKED -> "The agent is no longer waiting"
    Outcome.SIGNATURE_MISMATCH -> "The screen does not look like a prompt" + (r.detail?.let { " ($it)" } ?: "")
    Outcome.DIALOG_CHANGED -> "The dialog changed; nothing sent" + (r.detail?.let { " ($it)" } ?: "")
    Outcome.FAILED -> "Failed" + (r.detail?.let { ": $it" } ?: "")
    else -> r.outcome + (r.detail?.let { ": $it" } ?: "")
}

File: ios/Remotly/Views/ApprovalCard.swift
// Shown between the terminal and the key row while the watched pane is `blocked` (shared/design/DESIGN.md §4.5):
// the tool and the kind of dialog, the command or file, the agent's summary and question, and its choices as
// buttons in the dialog's own words. Tapping one asks the bridge to move the desktop cursor there and press Enter
// (`choose`), so the same card serves permission prompts, Claude's AskUserQuestion menus and pickers; the choice
// the desktop currently marks is highlighted. Esc is on the key row; "Deny with feedback…" / "Something else…"
// dismiss the dialog and type the reply. Without parsed details (trust dialogs, unknown layouts) the card falls
// back to the agent's key-map actions: Approve / Approve for session / Deny / Interrupt.
import FlowKit
import SwiftUI
import UIKit

// @MainActor: `result`, `inFlight` and `act` touch the main-actor AppModel outside `body`.
@MainActor
struct ApprovalCard: View {
    @Environment(AppModel.self) private var model
    let pane: Pane
    let promptId: String
    @State private var showFeedback = false
    @State private var feedback = ""

    private var result: ApprovalResult? { model.approvalResults[promptId] }
    private var inFlight: Bool { model.approvalInFlight == promptId }
    private var isChoice: Bool { pane.approval?.isChoice ?? false }
    private var sent: Bool { result?.outcome == .sent }
    private var quietTitle: String { isChoice ? "Something else…" : "Deny with feedback…" }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            head
            if let result, sent {
                Text(result.summaryText)
                    .font(.system(size: 14))
                    .foregroundStyle(Theme.fg2)
            } else {
                if let details = pane.approval {
                    parsed(details)
                } else {
                    fallback
                }
                if let result {
                    HStack(alignment: .firstTextBaseline, spacing: 12) {
                        Text(result.summaryText)
                            .font(.system(size: 14))
                            .foregroundStyle(Theme.fg2)
                            .lineLimit(2)
                        Spacer(minLength: 0)
                        if result.outcome == .signatureMismatch, model.lastApprovalAttempt[promptId] != nil {
                            Button("Send anyway") { model.retryForced(pane: pane.id, promptId: promptId) }
                                .buttonStyle(QuietButtonStyle(color: Theme.interactive))
                        }
                    }
                }
            }
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.panel)
        .overlay(alignment: .top) { Rectangle().fill(Theme.line).frame(height: 1) }
        .animation(.easeInOut(duration: 0.2), value: result?.outcome.rawValue)
        .animation(.easeInOut(duration: 0.2), value: inFlight)
        .sheet(isPresented: $showFeedback) { feedbackSheet }
    }

    // MARK: Head row

    private var head: some View {
        HStack(alignment: .firstTextBaseline) {
            if let result, !sent {
                Text("✕ Not sent")
                    .font(Theme.mono(13, bold: true))
                    .foregroundStyle(Theme.blocked)
                Spacer(minLength: 8)
                Text(result.outcome.rawValue.replacingOccurrences(of: "_", with: " "))
                    .font(Theme.mono(13))
                    .foregroundStyle(Theme.fg3)
            } else if sent {
                Text("✓ Answered from your phone")
                    .font(Theme.mono(13, bold: true))
                    .foregroundStyle(Theme.done)
                Spacer(minLength: 8)
                Text("sent")
                    .font(Theme.mono(13))
                    .foregroundStyle(Theme.fg3)
            } else {
                Text(pane.approval?.tool ?? (isChoice ? "Question" : "Approval"))
                    .font(Theme.mono(13, bold: true))
                    .foregroundStyle(Theme.accent)
                    .lineLimit(1)
                Spacer(minLength: 8)
                Text(inFlight ? "sending…" : (pane.approval == nil ? "prompt" : (isChoice ? "question" : "permission")))
                    .font(Theme.mono(13))
                    .foregroundStyle(Theme.fg3)
            }
        }
    }

    // MARK: Parsed dialog

    @ViewBuilder
    private func parsed(_ details: ApprovalDetails) -> some View {
        if let what = details.command ?? details.path {
            Text(what)
                .font(Theme.mono(12.5))
                .foregroundStyle(Theme.fg)
                .lineLimit(6)
                .textSelection(.enabled)
                .padding(.horizontal, 8)
                .padding(.vertical, 6)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(Theme.bg, in: RoundedRectangle(cornerRadius: 8))
                .overlay(RoundedRectangle(cornerRadius: 8).stroke(Theme.line, lineWidth: 1))
        }
        if let description = details.description, description != (details.command ?? details.path) {
            Text(description)
                .font(.system(size: 14))
                .foregroundStyle(Theme.fg2)
        }
        if !details.question.isEmpty {
            Text(details.question)
                .font(.system(size: 15, weight: .semibold))
                .foregroundStyle(Theme.fg)
        }
        VStack(spacing: 6) {
            ForEach(Array(details.options.enumerated()), id: \.offset) { index, option in
                let number = index + 1
                let marked = details.selected == number
                Button {
                    model.choose(pane: pane.id, promptId: promptId, option: number, label: option)
                } label: {
                    Text(option).lineLimit(3)
                }
                .buttonStyle(OptionButtonStyle(marked: marked))
                .accessibilityLabel("Option \(number): \(option)\(marked ? ", marked on the desktop" : "")")
            }
        }
        .disabled(inFlight)
        .opacity(inFlight ? 0.6 : 1)
        Button(quietTitle) { showFeedback = true }
            .buttonStyle(QuietButtonStyle())
            .disabled(inFlight)
    }

    // MARK: Nothing parsed

    @ViewBuilder
    private var fallback: some View {
        if let label = pane.stateLabel, !label.isEmpty {
            Text(label)
                .font(.system(size: 14))
                .foregroundStyle(Theme.fg2)
        }
        VStack(spacing: 6) {
            Button("Approve") { act(.approve) }.buttonStyle(OptionButtonStyle(marked: true))
            Button("Approve for session") { act(.approveSession) }.buttonStyle(OptionButtonStyle())
            Button("Deny") { act(.deny) }.buttonStyle(OptionButtonStyle())
            Button("Interrupt") { act(.interrupt) }.buttonStyle(OptionButtonStyle())
        }
        .disabled(inFlight)
        .opacity(inFlight ? 0.6 : 1)
        Button("Deny with feedback…") { showFeedback = true }
            .buttonStyle(QuietButtonStyle())
            .disabled(inFlight)
    }

    // MARK: Feedback sheet

    private var feedbackSheet: some View {
        NavigationStack {
            VStack(alignment: .leading, spacing: 10) {
                TextField("", text: $feedback, prompt: Text("What should it do instead?").foregroundStyle(Theme.fg3), axis: .vertical)
                    .lineLimit(3...10)
                    .font(.system(size: 15))
                    .foregroundStyle(Theme.fg)
                    .tint(Theme.interactive)
                    .padding(12)
                    .background(Theme.panel, in: RoundedRectangle(cornerRadius: 12))
                    .overlay(RoundedRectangle(cornerRadius: 12).stroke(Theme.line, lineWidth: 1))
                Text("Esc dismisses the dialog first, then your words are typed to the agent.")
                    .font(.system(size: 13))
                    .foregroundStyle(Theme.fg3)
                Spacer(minLength: 0)
            }
            .padding(16)
            .background(Theme.bg)
            .navigationBarTitleDisplayMode(.inline)
            .toolbarBackground(Theme.bg, for: .navigationBar)
            .toolbarBackground(.visible, for: .navigationBar)
            .toolbar {
                ToolbarItem(placement: .principal) {
                    Text(quietTitle).font(.system(size: 17, weight: .semibold)).foregroundStyle(Theme.fg)
                }
                ToolbarItem(placement: .cancellationAction) { Button("Cancel") { showFeedback = false } }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Send") {
                        act(.denyFeedback, feedback: feedback)
                        feedback = ""
                        showFeedback = false
                    }
                    .fontWeight(.semibold)
                    .disabled(feedback.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                }
            }
        }
        .tint(Theme.interactive)
        .presentationDetents([.medium])
        .presentationBackground(Theme.bg)
    }

    private func act(_ action: ApprovalAction, feedback: String? = nil) {
        model.approve(pane: pane.id, promptId: promptId, action: action, feedback: feedback)
    }
}
